lib/bap_future/bap_future.ml

```diff
@@ -0,0 +1,503 @@
+      | Wait _ -> invalid_arg "peek_exn: empty future"
+
+    module Monad = Monad.Make(struct
+        type nonrec 'a t = 'a t
+        let bind = bind
+        let map = `Custom map
+        let return = full
+      end)
+
+    module App = Applicative.Of_monad(struct
+        type nonrec 'a t = 'a t
+        include Monad
+      end)
+
+    module Args = Applicative.Make_args(struct
+        type nonrec 'a t = 'a t
+        include App
+      end)
+    include App
+    include Monad
+  end
+
+  type 'a future = 'a Future.t
+
+  type 'a signal = Signal of ('a -> unit)
+
+  module Promise = Future.P
+
+  type 'a promise = 'a Promise.t
+
+  module Signal = struct
+    type 'a t = 'a signal
+    let send (Signal send) x = send x
+
+    let repeat (Signal send) ~times x = 
+      let rec repeat' cnt = 
+        if cnt < times then (
+          send x; 
+          repeat' (cnt + 1)) in
+      repeat' 0
+
+  end
+
+  module Stream = struct
+    module Id = Int64
+    type id = Id.t
+
+
+    type 'a t = {
+      subs : (id -> 'a -> unit) Id.Table.t;
+      mutable last_id : id;
+      mutable on_subs: (id -> unit) list;
+      mutable on_unsubs: (id -> unit) list;
+      mutable waiters : (unit -> unit) list;
+    }
+
+    let add t f =
+      t.last_id <- Id.succ t.last_id;
+      Hashtbl.add_exn t.subs ~key:t.last_id ~data:f;
+      List.iter t.on_subs ~f:(fun f -> f t.last_id);
+      t.last_id
+
+    let subscribe t f =
+      let f id x = f x in
+      add t f
+
+    let watch s f = ignore (add s f)
+
+    let observe s f =
+      let f id x = f x in
+      ignore (add s f)
+
+    let unsubscribe t id =
+      Hashtbl.remove t.subs id;
+      List.iter t.on_unsubs ~f:(fun f -> f id)
+
+    let publish t event =
+      Hashtbl.iter t.subs ~f:(fun ~key:id ~data:notify ->
+          try notify id event with exn ->
+            unsubscribe t id;
+            raise exn)
+
+    let create () =
+      let stream = {
+        subs = Id.Table.create ();
+        last_id = Id.zero;
+        on_subs = [];
+        on_unsubs = [];
+        waiters = []
+      } in
+      stream, Signal (publish stream)
+
+    let wait t =
+      List.iter t.waiters ~f:(fun f -> f ())
+
+    let on_wait t f =
+      t.waiters <- f :: t.waiters
+
+    let on_subscribe t f =
+      t.on_subs <- f :: t.on_subs
+
+    let on_unsubscribe t f =
+      t.on_unsubs <- f :: t.on_unsubs
+
+    let is_closed t = Hashtbl.is_empty t.subs
+
+    let has_subscribers t = not (is_closed t)
+
+    let unfold' ~init ~f =
+      let stream,Signal push = create () in
+      let state = ref init in
+      let send () =
+        let q,s = f !state in
+        state := s;
+        Queue.iter q ~f:push in
+      stream, Signal send
+
+    let unfold ~init ~f =
+      let stream,Signal publish = create () in
+      let state = ref init in
+      let send () =
+        let x,s = f !state in
+        state := s;
+        publish x in
+      stream, Signal send
+
+    let unfold_until ~init ~f =
```

Comment:
the problem with the proposed approach with termination is that even after the stream has terminated and will never ever used,  it will be still hold alive and will be never garbage collected, thus leading to a memory leak. To fix this issue, I would propose to have two functions, one that does nothing, and another that actually sends data. Let `send` function will dispatch between them. Something like this:

```ocaml
    let do_nothing () = `finished in
    let do_publish () = ...  in
    let do_send = ref do_publish in
    let send () = 
      match !do_send () with
      | `finished -> do_send := do_nothing 
      | `continue -> () in
    ...
```